flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesClusterDescriptor.java

```diff
@@ -256,36 +244,51 @@ private String getWebMonitorAddress(Configuration configuration) throws Exceptio
                     flinkConfig.get(JobManagerOptions.PORT));
         }
 
+        final KubernetesJobManagerParameters kubernetesJobManagerParameters =
+                new KubernetesJobManagerParameters(flinkConfig, clusterSpecification);
+
+        final FlinkPod podTemplate =
+                kubernetesJobManagerParameters
+                        .getPodTemplateFilePath()
+                        .map(
+                                file ->
+                                        KubernetesUtils.loadPodFromTemplateFile(
+                                                client, file, Constants.MAIN_CONTAINER_NAME))
+                        .orElse(new FlinkPod.Builder().build());
+        final KubernetesJobManagerSpecification kubernetesJobManagerSpec =
+                KubernetesJobManagerFactory.buildKubernetesJobManagerSpecification(
+                        podTemplate, kubernetesJobManagerParameters);
+
+        client.createJobManagerComponent(kubernetesJobManagerSpec);
+
+        return createClusterClientProvider(clusterId);
+    }
+
+    private ClusterClientProvider<String> safelyDeployCluster(
+            SupplierWithException<ClusterClientProvider<String>, Exception> supplier)
+            throws ClusterDeploymentException {
         try {
-            final KubernetesJobManagerParameters kubernetesJobManagerParameters =
-                    new KubernetesJobManagerParameters(flinkConfig, clusterSpecification);
-
-            final FlinkPod podTemplate =
-                    kubernetesJobManagerParameters
-                            .getPodTemplateFilePath()
-                            .map(
-                                    file ->
-                                            KubernetesUtils.loadPodFromTemplateFile(
-                                                    client, file, Constants.MAIN_CONTAINER_NAME))
-                            .orElse(new FlinkPod.Builder().build());
-            final KubernetesJobManagerSpecification kubernetesJobManagerSpec =
-                    KubernetesJobManagerFactory.buildKubernetesJobManagerSpecification(
-                            podTemplate, kubernetesJobManagerParameters);
-
-            client.createJobManagerComponent(kubernetesJobManagerSpec);
-
-            return createClusterClientProvider(clusterId);
+
+            ClusterClientProvider<String> clusterClientProvider = supplier.get();
+
+            try (ClusterClient<String> clusterClient = clusterClientProvider.getClusterClient()) {
```

Comment:
I think that's the key part of this change except other code re-organization. But I don't understand the PR title for the following

- Is failing to get the cluster client same as failing to start that k8s cluster?
- Which code to kill the cluster --> just failing to deploy means killing?

Reply:
1. I think it means the failure when get cluster client failed, during the phase of deploy cluster. 
2. Failing to deploy means killing.

As I mentioned in [issue ](https://issues.apache.org/jira/browse/FLINK-24624?focusedCommentId=17434081&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17434081), This PR only solve the problem in case1. In session/application mode, we seems can not  totally ensure there are no left resource, because the deploy process is asynchronous, it may need to handle by the client.

Reply:
@cc13ny Thanks for your valuable comments.

@Aitozi This discussion make me to rethink that whether we really need to clean up the K8s resources when creating Flink client failed. Because the Flink cluster might be running normally.

Reply:
I prefer to handle the creating Flink client exception during the phase of `deployCluster`. 
I'm also ok to keep the current shape, it will not have too bad impact just with some resource object left.

Reply:
> Because the Flink cluster might be running normally.

I think the same. It turns out to be, if cleaning up the K8s resources is the best strategy or if it's possible to retry and re-create a client with the existing resource? Cleaning up is definitely the simplest way. It depends on how much and in which failing situations we would like to hand (e.g. re-create the client).

Reply:
Re-create the client probably does not make sense for FLINK-24624 since it will always fail due to permission issues.

After careful consideration, I lean to have more discussion and keep the current behavior. I still appreciate for @Aitozi 's work on this PR.

Reply:
Thanks for your comments @cc13ny @wangyang0918 , closing this PR